client/imgui-system.cpp

```diff
@@ -9,404 +9,419 @@
-		GLint success = 0;
-		glCompileShader(IMGUISystem::vertex_shader);
-		glGetShaderiv(IMGUISystem::vertex_shader, GL_COMPILE_STATUS, &success);
-		glCompileShader(IMGUISystem::fragment_shader);
-		glGetShaderiv(IMGUISystem::fragment_shader, GL_COMPILE_STATUS, &success);
-		glAttachShader(IMGUISystem::shader_program, IMGUISystem::vertex_shader);
-		glAttachShader(IMGUISystem::shader_program, IMGUISystem::fragment_shader);
-		glLinkProgram(IMGUISystem::shader_program);
-		GLint isLinked = 0;
-		glGetProgramiv(IMGUISystem::shader_program, GL_LINK_STATUS, &isLinked);
-
-		IMGUISystem::texture_attribute_location = glGetUniformLocation(IMGUISystem::shader_program, "Texture");
-		IMGUISystem::projmtx_attribute_location = glGetUniformLocation(IMGUISystem::shader_program, "ProjMtx");
-		IMGUISystem::position_attribute_location = glGetAttribLocation(IMGUISystem::shader_program, "Position");
-		IMGUISystem::uv_attribute_location = glGetAttribLocation(IMGUISystem::shader_program, "UV");
-		IMGUISystem::color_attribute_location = glGetAttribLocation(IMGUISystem::shader_program, "Color");
-
-		glGenBuffers(1, &IMGUISystem::vbo);
-		glGenBuffers(1, &IMGUISystem::ibo);
-
-		glGenVertexArrays(1, &IMGUISystem::vao);
-		glBindVertexArray(IMGUISystem::vao);
-		glBindBuffer(GL_ARRAY_BUFFER, IMGUISystem::vbo);
-		glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, IMGUISystem::ibo);
-		glEnableVertexAttribArray(IMGUISystem::position_attribute_location);
-		glEnableVertexAttribArray(IMGUISystem::uv_attribute_location);
-		glEnableVertexAttribArray(IMGUISystem::color_attribute_location);
-
-#define OFFSETOF(TYPE, ELEMENT) ((std::size_t)&(((TYPE *)0)->ELEMENT))
-		glVertexAttribPointer(IMGUISystem::position_attribute_location, 2, GL_FLOAT, GL_FALSE, sizeof(ImDrawVert), (GLvoid*)OFFSETOF(ImDrawVert, pos));
-		glVertexAttribPointer(IMGUISystem::uv_attribute_location, 2, GL_FLOAT, GL_FALSE, sizeof(ImDrawVert), (GLvoid*)OFFSETOF(ImDrawVert, uv));
-		glVertexAttribPointer(IMGUISystem::color_attribute_location, 4, GL_UNSIGNED_BYTE, GL_TRUE, sizeof(ImDrawVert), (GLvoid*)OFFSETOF(ImDrawVert, col));
+	this->RenderDrawLists(ImGui::GetDrawData());
+}
+
+void IMGUISystem::CreateDeviceObjects() {
+	const GLchar* _vertex_shader = "#version 330\n"
+								   "uniform mat4 ProjMtx;\n"
+								   "in vec2 Position;\n"
+								   "in vec2 UV;\n"
+								   "in vec4 Color;\n"
+								   "out vec2 Frag_UV;\n"
+								   "out vec4 Frag_Color;\n"
+								   "void main()\n"
+								   "{\n"
+								   "	Frag_UV = UV;\n"
+								   "	Frag_Color = Color;\n"
+								   "	gl_Position = ProjMtx * vec4(Position.xy,0,1);\n"
+								   "}\n";
+
+	const GLchar* _fragment_shader = "#version 330\n"
+									 "uniform sampler2D Texture;\n"
+									 "in vec2 Frag_UV;\n"
+									 "in vec4 Frag_Color;\n"
+									 "out vec4 Out_Color;\n"
+									 "void main()\n"
+									 "{\n"
+									 "	Out_Color = Frag_Color * texture( Texture, Frag_UV.st);\n"
+									 "}\n";
```

Comment:
I might recommend overriding this or putting it in a .inc file or something

Reply:
I think github is making this look a bit worse than it is, but you that's a good call out.